### PR TITLE
Refactor table reporter

### PR DESCRIPTION
**Pull request by fviernau**
Those are mainly preparations for further refactoring of the table reporter and addition of a new feature.
Please see individual commit messages for the details.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1079)

### CODE REVIEW

File: reporter/src/main/kotlin/reporters/ExcelReporter.kt [hunk 1]
```diff
@@ -76,7 +83,10 @@ class ExcelReporter : TableReporter() {
 
     private lateinit var creationHelper: CreationHelper
 
-    override fun generateReport(tabularScanRecord: TabularScanRecord, outputDir: File): File {
+    override fun generateReport(ortResult: OrtResult, resolutionProvider: ResolutionProvider,
+                                copyrightGarbage: CopyrightGarbage, outputDir: File, postProcessingScript: String?)
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/ExcelReporter.kt]: Put the parameters on separate lines:
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: reporter/src/main/kotlin/reporters/ExcelReporter.kt [hunk 2]
```diff
@@ -181,7 +191,7 @@ class ExcelReporter : TableReporter() {
         (0..1).forEach { sheet.autoSizeColumn(it) }
     }
 
-    private fun createSummarySheet(workbook: XSSFWorkbook, name: String, file: String, table: SummaryTable,
+    private fun createSummarySheet(workbook: XSSFWorkbook, name: String, file: String, table: TabularModelMapper.SummaryTable,
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/ExcelReporter.kt]: Put the parameters on separate lines:
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt [hunk 3]
```diff
@@ -22,16 +22,25 @@ package com.here.ort.reporter.reporters
...
 import com.here.ort.utils.log
 
 import java.io.File
 import java.util.SortedMap
 
-class StaticHtmlReporter : TableReporter() {
-    override fun generateReport(tabularScanRecord: TabularScanRecord, outputDir: File): File {
+class StaticHtmlReporter : Reporter() {
+    override fun generateReport(ortResult: OrtResult, resolutionProvider: ResolutionProvider,
+                            copyrightGarbage: CopyrightGarbage, outputDir: File, postProcessingScript: String?)
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/StaticHtmlReporter.kt]: Put the parameters on separate lines:
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: reporter/src/main/kotlin/reporters/ExcelReporter.kt [hunk 4]
```diff
@@ -20,10 +20,17 @@
 package com.here.ort.reporter.reporters
 
 import ch.frankel.slf4k.*
+import com.here.ort.model.OrtResult
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/ExcelReporter.kt]: nit: Empty line above, no empty line below.

File: reporter/src/main/kotlin/reporters/ExcelReporter.kt [hunk 5]
```diff
@@ -20,10 +20,17 @@
...
 
 import com.here.ort.model.ScanRecord
 import com.here.ort.model.VcsInfo
+import com.here.ort.model.config.CopyrightGarbage
 import com.here.ort.model.config.ProjectExclude
+
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/ExcelReporter.kt]: nit: Remove empty line.

File: reporter/src/main/kotlin/reporters/ExcelReporter.kt [hunk 6]
```diff
@@ -20,10 +20,17 @@
...
+import com.here.ort.reporter.Reporter
+import com.here.ort.reporter.ResolutionProvider
+import com.here.ort.reporter.reporters.TabularModelMapper.ProjectTable
+
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/ExcelReporter.kt]: nit: Remove empty line.

File: reporter/src/main/kotlin/reporters/TableReporter.kt [hunk 7]
```diff
@@ -219,8 +219,8 @@ abstract class TableReporter : Reporter() {
     }
 
     data class ResolvableError(
-            val error: OrtIssue,
-            val resolutions: List<ErrorResolution>
+            private val error: OrtIssue,
+            private val resolutions: List<ErrorResolution>
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/TableReporter.kt]: Now that you are refactoring this class, you could also rename it to `ResolvableIssue`.

File: reporter/src/main/kotlin/reporters/TableReporter.kt [hunk 8]
```diff
@@ -229,6 +229,8 @@ abstract class TableReporter : Reporter() {
                         append(resolutions.joinToString(prefix = "\nResolved by: ") { "${it.reason} - ${it.comment}" })
                     }
                 }
+
+        fun isResolved() = !resolutions.isEmpty()
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/TableReporter.kt]: Use `isNotEmpty()`. Also make it `val resolved` instead, as `resolutions` is immutable.

File: reporter/src/main/kotlin/reporters/TableReporter.kt [hunk 9]
```diff
@@ -222,7 +222,7 @@ abstract class TableReporter : Reporter() {
             val error: OrtIssue,
             val resolutions: List<ErrorResolution>
     ) {
-        override fun toString() =
+        fun getDescription(): String  =
```
Comment (mnonnenmacher) [inline reporter/src/main/kotlin/reporters/TableReporter.kt]: The function always returns the same result, so if you do not override `toString()` anymore you could also make it a `val`.

File: reporter/src/main/kotlin/reporters/ReportTableModel.kt [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        }
+    }
+
+    data class ResolvableError(
+            private val error: OrtIssue,
+            private val resolutions: List<ErrorResolution>
+    ) {
+        fun getDescription() = buildString {
+            append(error)
+            if (resolutions.isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply reporter/src/main/kotlin/reporters/ReportTableModel.kt]: Yes.

**Pull request merged**